### ISSUE TITLE
Egress to external resources broken on all non host-network pods for a few nodes every time the cluster is scaled up

### ISSUE THREAD

**Issue by htoo97** [cilium/cilium#19477]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

TLDR is that on the bad nodes, all (non host-network) pods cannot make egress requests to external resources (such as a random IP belonging to google.com). Talking to cluster-internal IPs all work. Somehow the TCP 3-way handshake was never completed and always failed at the ACK, which had its source pod IP not being masqueraded properly as the node (more info from tcpdump outputs below). Interestingly, churning and restarting the cilium-agent container always fixes it.

### Background
In a cluster set up with kops, we are using Cilium with external ETCD and using the default overlay network routing. Not much configuration is updated, and the default kube-proxy replacement mode of probe is being used. We did not run into this issue in the past few months we have been doing R&D but as we got closer to launching, we added autoscaling schedules (just the AWS ones) to scale up and down our fleet daily. The first time this happened, we were testing scaling up the cluster from ~200 nodes close to about ~950 nodes, and noticed this issue in about 11 nodes. Subsequent times that happened, we've scaled up the fleet by about 130 nodes and got these 1-2 bad nodes at a time. This also only happens to new nodes and never to already running nodes. In this cluster, we are also disabling any cilium policy enforcement (ie. all traffic is allowed).

In the following examples, I will be using `142.251.111.102` (a random google.com IP) as the destination, `<POD_IP>` as the redacted pod IP and `<NODE_IP>` as the redacted node IP.

### Issue
On any pod's network namespace on the bad node, raw HTTP requests to external IPs always hangs. The same thing happens with FQDNs too but its DNS-resolved IPs just get retried to the same effect.
```
/app# curl -vv 142.251.111.102
* Rebuilt URL to: 142.251.111.102/
*   Trying 142.251.111.102...
* TCP_NODELAY set
* Connected to 142.251.111.102 (142.251.111.102) port 80 (#0)
> GET / HTTP/1.1
> Host: 142.251.111.102
> User-Agent: curl/7.58.0
> Accept: */*
> 
^C
```

### Debugging with Hubble
On the cilium agent pod, after triggering the curl from a pod, doing a hubble observe results in the following. Notice that there is no return packet to the pod from the destination IP.
```
/home/cilium# hubble observe --follow --ip=<POD_IP> --debug | grep 142.251.111.102
Apr 12 20:44:14.830: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: SYN)
Apr 12 20:44:14.839: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: ACK)
Apr 12 20:44:14.839: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: ACK, PSH)
Apr 12 20:44:16.433: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: ACK, FIN)
Apr 12 20:44:16.567: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: ACK, FIN, PSH)
Apr 12 20:44:18.295: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: ACK, FIN, PSH)
Apr 12 20:44:21.212: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: ACK)
Apr 12 20:44:21.879: NS/POD_NAME:45234 -> 142.251.111.102:80 to-stack FORWARDED (TCP Flags: ACK, FIN, PSH)
```

### Debugging with tcpdump
Repeating the same request above on the node and running tcpdump on the node itself results in the following. The tcpdump on the pod level looks practically similar when comparing pods on a good vs bad node. Please refer to the in-line comments for the observations so far about the TCP 3-way handshake.
```
# (1) The initial SYN packet from pod -> external IP
# - The 2 packets have the same exact identification 64430 but 
#    everything else except the TTL is exactly the same.
# - The TTL is 2 hops less on the packet with the node IP as the source. 
#    I've also noticed that this is the same TTL that shows up on reproducing 
#    on a good node, so I'm assuming 2 hops are consumed for cilium related things to do NAT-ing?
/home/ubuntu# tcpdump -n -i any host 142.251.111.102 -vv -S
00:57:41.312018 IP (tos 0x0, ttl 64, id 64430, offset 0, flags [DF], proto TCP (6), length 60)
    <POD_IP>.44188 > 142.251.111.102.80: Flags [S], cksum 0x7376 (incorrect -> 0x8aa3), seq 2938793639, win 65535, options [mss 8911,sackOK,TS val 415844720 ecr 0,nop,wscale 9], length 0
00:57:41.312083 IP (tos 0x0, ttl 62, id 64430, offset 0, flags [DF], proto TCP (6), length 60)
    <NODE_IP>.44188 > 142.251.111.102.80: Flags [S], cksum 0xe03b (incorrect -> 0x1dde), seq 2938793639, win 65535, options [mss 8911,sackOK,TS val 415844720 ecr 0,nop,wscale 9], length 0

# (2) The SYN-ACK reply from the external IP
# - The reply comes back to the node IP. 
# - The seq and ack numbers are as expected.
00:57:41.391941 IP (tos 0x0, ttl 47, id 58105, offset 0, flags [none], proto TCP (6), length 60)
    142.251.111.102.80 > <NODE_IP>.44188: Flags [S.], cksum 0x86f9 (correct), seq 2450758419, ack 2938793640, win 65535, options [mss 1430,sackOK,TS val 2543841094 ecr 415844720,nop,wscale 8], length 0

# (3) The ACK back from pod -> external IP
# - The source is now the pod IP, instead of using the masqueraded node IP,
#   which is what's observed on good nodes.
# - The seq and acks still follow from above.
00:57:41.391981 IP (tos 0x0, ttl 64, id 64431, offset 0, flags [DF], proto TCP (6), length 52)
    <POD_IP>.44188 > 142.251.111.102.80: Flags [.], cksum 0x736e (incorrect -> 0x219e), seq 2938793640, ack 2450758420, win 128, options [nop,nop,TS val 415844800 ecr 2543841094], length 0

# The pod attempts to make a request.
00:57:41.392047 IP (tos 0x0, ttl 64, id 64432, offset 0, flags [DF], proto TCP (6), length 131)
    <POD_IP>.44188 > 142.251.111.102.80: Flags [P.], cksum 0x73bd (incorrect -> 0x6105), seq 2938793640:2938793719, ack 2450758420, win 128, options [nop,nop,TS val 415844800 ecr 2543841094], length 79: HTTP, length: 79
	GET / HTTP/1.1
	Host: 142.251.111.102
	User-Agent: curl/7.58.0
	Accept: */*
	
# The external IP now resends the SYN-ACK, with the same exact seq and ack numbers as (2).
# - This suggests that the external IP never received the (3)rd ACK.
00:57:41.698219 IP (tos 0x0, ttl 47, id 58229, offset 0, flags [none], proto TCP (6), length 60)
    142.251.111.102.80 > <NODE_IP>.44188: Flags [S.], cksum 0x85c7 (correct), seq 2450758419, ack 2938793640, win 65535, options [mss 1430,sackOK,TS val 2543841400 ecr 415844720,nop,wscale 8], length 0

# The pod then resends the ACK to the external IP. This repeats and the request hangs.
00:57:41.698233 IP (tos 0x0, ttl 64, id 64433, offset 0, flags [DF], proto TCP (6), length 52)
    <POD_IP>.44188 > 142.251.111.102.80: Flags [.], cksum 0x736e (incorrect -> 0x201d), seq 2938793719, ack 2450758420, win 128, options [nop,nop,TS val 415845106 ecr 2543841094], length 0
00:57:41.702379 IP (tos 0x0, ttl 64, id 64434, offset 0, flags [DF], proto TCP (6), length 131)
    <POD_IP>.44188 > 142.251.111.102.80: Flags [P.], cksum 0x73bd (incorrect -> 0x5fcf), seq 2938793640:2938793719, ack 2450758420, win 128, options [nop,nop,TS val 415845110 ecr 2543841094], length 79: HTTP, length: 79
	GET / HTTP/1.1
	Host: 142.251.111.102
	User-Agent: curl/7.58.0
	Accept: */*
```
One interesting thing about the TTLs above is that for all the packets that are wrongly using the pod IP as the source IP instead of the masqueraded node IP, the TTLs are 64, 2 hops more than 62 that I observe on corresponding packets on the good nodes.

I'm pretty new to Cilium myself, so I wanted to see if anyone else has run into this issue as well, or if anyone can point me to code that I can look closer in. Would appreciate any help!

### Cilium Version

```
Client: 1.9.13 f2160fc 2022-02-23T13:03:31-08:00 go version go1.15.15 linux/amd64
Daemon: 1.9.13 f2160fc 2022-02-23T13:03:31-08:00 go version go1.15.15 linux/amd64
```

### Kernel Version

```
Linux <HOSTNAME> 5.11.0-1021-aws #22~20.04.2-Ubuntu SMP Wed Oct 27 21:27:13 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

### Kubernetes Version

1.15.12

### Sysdump

I lost access to a borked node, but will paste the sysdump from one when it happens again. It hasn't entirely been deterministically reproducible.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by borkmann** [comment]
> Client: 1.9.13 f2160fc 2022-02-23T13:03:31-08:00 go version go1.15.15 linux/amd64
> Daemon: 1.9.13 f2160fc 2022-02-23T13:03:31-08:00 go version go1.15.15 linux/amd64

Did you also test if the issue exist in latest 1.11 stable release?

**Comment by htoo97** [comment]
I think the version of Cilium that I have does not have the sysdump subcommand yet. I followed the instructions [here](https://docs.cilium.io/en/v1.9/operations/troubleshooting/#automatic-log-state-collection). Will DM you the zip on Cilium Slack.

This [debuginfo.txt](https://github.com/cilium/cilium/files/8544672/debuginfo.txt) was also obtained from running `cilium debuginfo` on the bad node's agent.

> Did you also test if the issue exist in latest 1.11 stable release?
We can't go above 1.9 because of k8s version compatibility issues. We have plans to upgrade to 1.18 which would open up using a later version. I just wanted to post first to gather if this is a known bug/issue. If not, I can report back after the k8s upgrade (using the same cilium version) and then after upgrading cilium itself as well after.
